src/AudioCaptureImpl_WASAPI.cpp

```diff
@@ -1,25 +1,631 @@
+    {
+        poco_trace(_logger, "Releasing audio client.");
+        _audioClient->Release();
+        poco_trace(_logger, "Audio client released.");
+        _audioClient = nullptr;
+    }
+
+    if (device)
+    {
+        poco_trace(_logger, "Releasing audio device.");
+        device->Release();
+    }
+}
+
+void AudioCaptureImpl::CaptureThread()
+{
+    poco_debug(_logger, "Audio capture thread starting.");
+
+    HRESULT result = CoInitializeEx(nullptr, COINIT_MULTITHREADED);
+
+    if (FAILED(result))
+    {
+        poco_error_f1(_logger, "CoInitializeEx() failed: result = 0x%08?x", result);
+        throw std::bad_alloc();
+    }
+
+    IMMDeviceEnumerator* enumerator{GetDeviceEnumerator()};
+
+    if (enumerator == nullptr)
+    {
+        CoUninitialize();
+        throw std::bad_alloc();
+    }
+
+    poco_trace(_logger, "Registering device callbacks.");
+    enumerator->RegisterEndpointNotificationCallback(this);
+
+    do
+    {
+        _restartCapturing = false;
+
+        auto devices{GetAudioDeviceList(enumerator)};
+        bool useLoopback{true};
+        std::string deviceName;
+
+        IMMDevice* device{nullptr};
+
+        if (_currentAudioDeviceIndex == -1 || _currentAudioDeviceIndex >= devices.size())
+        {
+            // Get the default render endpoint for opening it as a loopback device.
+            result = enumerator->GetDefaultAudioEndpoint(eRender, eConsole, &device);
+            if (FAILED(result))
+            {
+                poco_error_f1(_logger, "IMMDeviceEnumerator::GetDefaultAudioEndpoint failed: result = 0x%08?x", result);
+                break;
+            }
+
+            deviceName = _defaultDeviceName;
+        }
+        else
+        {
+            // Get a device by its ID according to the currently selected index.
+            useLoopback = devices.at(_currentAudioDeviceIndex).IsRenderDevice();
+            deviceName = devices.at(_currentAudioDeviceIndex).FriendlyName();
+            result = enumerator->GetDevice(devices.at(_currentAudioDeviceIndex).DeviceId(), &device);
+
+            if (FAILED(result))
+            {
+                poco_error_f1(_logger, "IMMDeviceEnumerator::GetDevice failed: result = 0x%08?x", result);
+                break;
+            }
+        }
+
+        LPWSTR deviceID{nullptr};
+        result = device->GetId(&deviceID);
+        if (FAILED(result))
+        {
+            poco_error_f1(_logger, "IMMDevice::GetId failed: result = 0x%08?x", result);
+        }
+        else
+        {
+            _currentCaptureDeviceId = UnicodeToString(deviceID);
+        }
+
+        if (!OpenAudioDevice(device, useLoopback))
+        {
+            _isCapturing = false;
+        }
+
+        poco_information_f3(_logger, "Audio device opened: %s (channels: %hu, loopback: %b)", deviceName, _channels, useLoopback);
+
+        while (_isCapturing && !_restartCapturing)
+        {
+            try
+            {
+                _fillBufferEvent.tryWait(500);
```

Comment:
Ideally, that timeout should never trigger, it's just a safeguard to avoid a deadlock in case something else breaks and the event isn't properly signaled.